smartlearner/optimizers/optimizer.py

```diff
@@ -12,12 +12,22 @@ def __init__(self, loss):
         self._update_rules = []
         self._param_modifiers = []
 
+        self.graph_updates = OrderedDict()
+        self._directions = None
+
     def append_update_rule(self, update_rule):
         self._update_rules.append(update_rule)
 
     def append_param_modifier(self, param_modifier):
         self._param_modifiers.append(param_modifier)
 
+    @property
+    def directions(self):
+        if self._directions is None:
```

Comment:
We'll need to make sure the graph never changes (from a graph modification, for example). Otherwise this check will create headaches for us later.